lib/daffy_lib/services/key_management_service.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+require 'logger'
+
+class DaffyLib::KeyManagementService
+  class KeyManagementServiceException < StandardError; end
+  class InvalidParameterException < KeyManagementServiceException; end
+  class KeyGenerationException < KeyManagementServiceException; end
+  class KeyCreateException < KeyManagementServiceException; end
+  class KeyRetrieveException < KeyManagementServiceException; end
+
+  KEY_MANAGEMENT_SERVICE_CACHE_NAMESPACE = :key_management_service
+  KEY_VERSION = 'KeyManagementService::V1'
+
+  RECORD_NOT_UNIQUE_REGEX = /has already been taken/.freeze
+
+  def initialize(partition_guid, expires_in, cmk_key_id)
+    raise InvalidParameterException unless partition_guid.present? && expires_in.present? && cmk_key_id.present?
+
+    @partition_guid = partition_guid
+    @expire_in = expires_in
+    @cmk_key_id = cmk_key_id
+  end
+
+  def self.encryption_key_epoch(datetime)
+    datetime.utc.beginning_of_year
+  end
+
+  def find_or_create_encryption_key(encryption_epoch)
+    find(encryption_epoch) || create(encryption_epoch)
```

Comment:
This may fail in high concurrency (if you add the correct unique key). Look at create_or_find_by

Reply:
Fair point.  Changed it to use `create_or_find_by`.

Reply:
Upon discussion and reviewing with @bcarr092 , this is already built-in to how our methods are designed.  The `create` has a rescue block for `ActiveRecord::RecordNotUnique` that will call `find`.